compiler/noirc_evaluator/src/ssa/ir/dfg.rs

```diff
@@ -21,6 +22,69 @@ use serde::{Deserialize, Serialize};
 use serde_with::serde_as;
 use serde_with::DisplayFromStr;
 
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub(crate) struct CallStackId(u32);
+
+impl CallStackId {
+    pub(crate) fn root() -> Self {
+        Self::new(0)
+    }
+
+    fn new(id: usize) -> Self {
+        Self(id as u32)
+    }
+
+    pub(crate) fn index(&self) -> usize {
+        self.0 as usize
+    }
+
+    pub(crate) fn is_root(&self) -> bool {
+        self.0 == 0
+    }
+
+    /// Construct a CallStack from a CallStackId
+    fn get_call_stack(&self, locations: &[LocationNode]) -> CallStack {
+        let mut call_stack = List::new();
+        let mut current_location = *self;
+        while let Some(parent) = locations[current_location.index()].parent {
+            call_stack.push_back(locations[current_location.index()].value);
+            current_location = parent;
+        }
+        call_stack
+    }
+
+    /// Adds a location to the call stack
+    fn add_child(&self, location: Location, locations: &mut Vec<LocationNode>) -> CallStackId {
+        if let Some(result) = locations[self.index()]
+            .children
+            .iter()
+            .find(|child| locations[child.index()].value == location)
```

Comment:
If we can remove this .iter .find that may fix the performance concerns

Reply:
I have modified it with a heuristic that works well for regression_4709.

I have a more complicated version that I'd rather do as a follow-up PR because I am not sure it would be useful since the heuristic does not degrade the memory improvement at all, so using an exact version for the search would only get less performance.

Reply:
Sounds good to me